### PR TITLE
fix: handle pending authorization status in GET order

### PR DESCRIPTION
**Pull request by pietro-tota**
#### List of Changes
Handle GET orders NPG `PENDING` instead of not authorized.
If for a transaction is received pending state the refund is retried and the GET orders api call performed again to retrieve the latest status.

The first GET orders api call is postponed for a configurable timeout (10 minutes as per initial configuration) for those transaction stuck in AUTHORIZATION_REQUESTED status from the authorization requested timestamp: this delay in elaborating refund process for NPG transactions is required in order to make NPG close transaction and reach a final status so that get orders return OK/KO outcome and refund can be process accordingly.

In case a PENDING status is received after that timeout another attempt for refund is performed in order to retrieve a final status as much time as possible.
GET orders is retried also for 5xx/timeout errors and stop for 4xx or invalid response statuses received such as missing fields or unexpected status (transaction already refunded).
For any error during process an event is written to dead letter queue (GET orders receiving PENDING for all retries/5xx/timeouts) or in case of blocking errors (4xx/invalid response received for missing fields/unexpected status) 

#### Motivation and Context
This modification is needed in order to handle NPG "pending" status while performing refund, performing others api calls.

There is an open point with this PR:  if for a transaction an error is received performed GET orders on the first attempt and this error is then recovered receiving an not authorized status the transaction will remain in REFUND_ERROR status.
This will be covered reviewing this flow (splitting GET order from refund process or introducing a new final status) to cover those cases, checking also the production cases after deploying this fix

#### How Has This Been Tested?
junit tests + local tests with eCommerce local 

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt [hunk 1]
```diff
@@ -1009,3 +1066,48 @@ fun isRefundableCheckRequired(tx: BaseTransaction): Boolean =
       setOf(TransactionStatusDto.AUTHORIZATION_COMPLETED, TransactionStatusDto.CLOSURE_REQUESTED)
         .contains(tx.status) && !wasAuthorizationDenied(tx)
   }
+
+fun <T> postponeRefundProcessingRequest(
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt:1070]: This doesn't actually `postpone` the request, what about:
```suggestion
fun <T> computeRefundProcessingRequestDelay(
```
?

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt [hunk 2]
```diff
@@ -1009,3 +1066,48 @@ fun isRefundableCheckRequired(tx: BaseTransaction): Boolean =
...
+      .next()
+      .map { ZonedDateTime.parse(it.creationDate) }
+      .switchIfEmpty(
+        Mono.error(
+          RuntimeException("No auth requested event found for a transaction to be refunded")))
+
+  if (authOutcome != null) {
+    return Mono.just(Duration.ZERO)
+  }
+  // actually refund to be postponed only for NPG gateway
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt:1093]: Wording
```suggestion
  // refund can be postponed only for NPG gateway
```

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/TransactionExpirationQueueConsumer.kt [hunk 3]
```diff
@@ -155,14 +155,35 @@ class TransactionExpirationQueueConsumer(
...
           val tracingInfo = queueEvent.fold({ it.tracingInfo }, { it.tracingInfo })
-          requestRefundTransaction(
-            tx, // transaction
-            transactionsRefundedEventStoreRepository,
-            transactionsViewRepository,
-            npgService,
-            tracingInfo,
-            refundRequestedAsyncClient,
-            Duration.ofSeconds(transientQueueTTLSeconds.toLong()))
+          val timeoutForRefund =
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/TransactionExpirationQueueConsumer.kt:158]: What about delay?
```suggestion
          val delayForRefund =
```

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt [hunk 4]
```diff
@@ -465,35 +465,29 @@ fun refundTransaction(
...
-                      else ->
-                        throw IllegalArgumentException(
-                          "Cannot determine touch point: [$it] for redirect transaction refund")
-                    }
-                  },
-                pspTransactionId = transactionAuthorizationRequestData.authorizationRequestId,
-                paymentTypeCode = transactionAuthorizationRequestData.paymentTypeCode,
-                pspId = transactionAuthorizationRequestData.pspId)
-            }
+          refundTransactionNPG(
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt:470]: It seems that the call to `appendNpgRefundRequestedEventIfNeeded` is gone, is it ok?

Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt:470]: yes, the refund transaction code is called by refund requested/refund error event handlers. Those handlers already require a transaction to be in refund requested / refund error event so this code will never write this event.

**Review by infantesimone — APPROVED** [review]

**Pull request merged**